### ISSUE TITLE
UWSGI container appears to be missing settings.py file

### ISSUE THREAD

**Issue by surewould** [DefectDojo/django-DefectDojo#1036]
**Bug description**
uwsgi container appears to be missing the settings.py file and fails to start the application

**Steps to reproduce**
1. Follow docker quick start... docker-compose up
2. Scroll down to initial uwsgi_1 status

```
uwsgi_1         | Traceback (most recent call last):
uwsgi_1         |   File "./dojo/wsgi.py", line 29, in <module>
uwsgi_1         |     application = get_wsgi_application()
uwsgi_1         |   File "/usr/local/lib/python2.7/site-packages/django/core/wsgi.py", line 13, in get_wsgi_application
uwsgi_1         |     django.setup(set_prefix=False)
uwsgi_1         |   File "/usr/local/lib/python2.7/site-packages/django/__init__.py", line 22, in setup
uwsgi_1         |     configure_logging(settings.LOGGING_CONFIG, settings.LOGGING)
uwsgi_1         |   File "/usr/local/lib/python2.7/site-packages/django/conf/__init__.py", line 56, in __getattr__
uwsgi_1         |     self._setup(name)
uwsgi_1         |   File "/usr/local/lib/python2.7/site-packages/django/conf/__init__.py", line 41, in _setup
uwsgi_1         |     self._wrapped = Settings(settings_module)
uwsgi_1         |   File "/usr/local/lib/python2.7/site-packages/django/conf/__init__.py", line 110, in __init__
uwsgi_1         |     mod = importlib.import_module(self.SETTINGS_MODULE)
uwsgi_1         |   File "/usr/local/lib/python2.7/importlib/__init__.py", line 37, in import_module
uwsgi_1         |     __import__(name)
uwsgi_1         | ImportError: No module named settings
```

settings.py appears to be missing

```
root@a5e158be0e3b:/app/dojo/settings# ls
__init__.py  settings.dist.py  template-env  unittest.py
root@a5e158be0e3b:/app/dojo/settings# 
```

**Expected behavior**
I expected the application to load as per the quick start.

**Deployment method** *(select with an `X`)*
- [ ] Kubernetes
- [X] Docker
- [ ] setup.bash / legacy-setup.bash

**Environment information**
 - Operating System: Manjaro 18.0.4
 - DefectDojo version: d957356: Merge pull request #1022 from DefectDojo/dev [2019-04-08 15:09:21 -0400]

**Comment by surewould** [comment]
I may have been premature. This looks like it is well explored below.

- [issues/1026](https://github.com/DefectDojo/django-DefectDojo/issues/1026)
- [pull/1028](https://github.com/DefectDojo/django-DefectDojo/pull/1028)

**Comment by ptrovatelli** [comment]
Hi @surewould ,
Yeap this is a duplicate of https://github.com/DefectDojo/django-DefectDojo/issues/1026 . There is a PR there with a workaround that changes docker-compose configuration files. However now that you have spotted that the error is due to this missing file, it might be possible to do a more simple fix and keep the docker-compose override thing - not sure it's the best way to go though, see discussions in the PR.

**Comment by surewould** [comment]
I found a fix.

**TL;DR** 'cp dojo/settings/settings.dist.py dojo/settings/settings.py' from django-DefectDojo folder and rebuild.

Steps taken:

I deleted all of the old images out of paranoia. Noticing my docker version was oh so slightly behind the latest package, I updated that. Reboot, just to be sure. Yes, deep admin magic. Lol.

I proceeded to follow the docker quick start instructions

```
$ git clone https://github.com/DefectDojo/django-DefectDojo
$ cd django-DefectDojo
$ docker-compose up
```

I ended up with a borken install consistent with what happened before.

However, I kept a close eye on the docker build process and spotted this.

```
Step 9/14 : RUN   cp dojo/settings/settings.dist.py dojo/settings/settings.py
```

Rooting around in the docker-compose.yml file I spotted this.

```
  uwsgi:
    image: defectdojo/defectdojo-django:latest
    build: 
      context: ./
      dockerfile: Dockerfile.django
```

I took a look at Dockerfile.django and saw this.

```
COPY dojo/ ./dojo/
# Legacy installs need the modified settings.py, do not remove!
RUN \
  cp dojo/settings/settings.dist.py dojo/settings/settings.py
```

That's weird. It looks like they are very specifically copying over settings.dist.py to settings.py and warning us not to remove that line, but it doesn't end up in final image. See original post.

So I did what any lazy admin would do.

```
$ ^C
$ docker-compose down --volumes
$ cp dojo/settings/settings.dist.py dojo/settings/settings.py
$ docker-compose up --build
```

It worked!

I don't know what is up with that line in Dockerfile.django. It looks right to me, but it clearly isn't making the copy happen.

**Comment by ptrovatelli** [comment]
@surewould nice finding! however I have to say I don't understand how working on your local git clone would have an impact on "docker-compose up". Did you rebuild the uwsgi  image after copying settings.dist.py to settings.py? 
From what I understand (@aaronweaver correct my if i'm wrong) since we have introduced the override file this I what we wanted to do: 
- "docker-compose up" will use docker-compose.yml and docker-compose.override.yml as this is the default behavior.  docker-compose.override.yml is supposed to override the uwsgi service configuration so that we'll pull defectdojo/defectdojo-django:latest (which is the latest locally built image with "docker build..." or the dockerhub one if none was locally built) and work on that in the container
- "docker-compose -f docker-compose.yml up" will force the use of docker-compose.yml (and thus ignore the override file). With this, we're working directly on the local git cloned files, so that we can develop and test without building the image after every change

If you haven't rebuilt the image, it means that the override is not fully overriding but we end up with a mix between both local files and the image, or maybe just local files. Docker-compose documentation does state that the override is not overriding but merging configuration, which is why in my PR I completely separated the "dev process" (work on local git cloned files) from the "release process" (work on the image) to make things more predictible. See doc at https://docs.docker.com/compose/extends/
There is actually something about overrides with "build" on one part and "image" on the other but it's only about version 1 file (we use version 3) (see "build and image in Compose file version 1"). It says that in that case "build" wins. maybe that's what's happening so we're actually always building from git cloned files, and never use a pre-built image. 
If that's the case it's  problematic, we need to be able to run the app from an existing image

**Comment by aaronweaver** [comment]
The script I wrote for dev should copy it over. I need to double check that it was copied over, I was pretty sure that was in that entrypoint script. Yes, for legacy reasons we can't have that in place.

**Comment by ptrovatelli** [comment]
it's there in entrypoint-uwsgi-dev.sh called by uwsgi in docker-compose.yml but not in entrypoint-uwsgi.sh called by the override. We're probably calling the override entrypoint with "docker-compose up"

**Comment by surewould** [comment]
@ptrovatelli Thank you! Full disclosure. I have no idea what I'm doing with docker... yet. ¯\_(ツ)_/¯

> Did you rebuild the uwsgi image after copying settings.dist.py to settings.py?

Yes, and thank you for the clarification of what is going on here. I'd be happy to let the experts take it from here, sit back, watch, and try to learn something.

**Comment by ptrovatelli** [comment]
@surewould right sorry I hadn't see this "--build" after your docker-compose up. I usually use 
`docker build -t defectdojo/defectdojo-django -f Dockerfile.django .`

I've just done a simple test to make sure about this configuration merge; from dev branch: 
```
git reset --hard
git status
# (manually remove all unversionned files)
git pull
docker-compose down --volumes
docker-compose up
```

This reproduces the "ImportError: No module named settings"

Now fix with: 
```
cp dojo/settings/settings.dist.py dojo/settings/settings.py
docker-compose up
```

Tada it works, we don't even need to rebuild the image. 
Let's check the volumes: 
```
$ docker inspect -f "{{ .Mounts }}" django-defectdojo_uwsgi_1
[{bind  /home/pascal/securite/django-DefectDojo /app  rw true rprivate}]
```

This is the volume from the uwsgi service in docker-compose.yml. 

Now if we sabotage a little our entrypoints: 
```
docker-compose.yml:
entrypoint: ['/wait-for-it.sh', 'mysql:3306', '-t', '30', '--', '/entrypoint-uwsgi-not-override.sh']
docker-compose.override.yml: 
entrypoint: ['/wait-for-it.sh', 'mysql:3306', '-t', '30', '--', '/entrypoint-uwsgi_override.sh']

```
And restart:
```
docker-compose down --volumes
docker-compose up
 docker logs django-defectdojo_uwsgi_1
wait-for-it.sh: waiting 30 seconds for mysql:3306
wait-for-it.sh: mysql:3306 is available after 14 seconds
/wait-for-it.sh: line 158: /entrypoint-uwsgi_override.sh: No such file or directory
```

We can see that the override entrypoint is called. 
So we have the volume from the docker-compose.yml uwsgi service, but the entrypoint from docker-compose.override.yml uwsgi service. Isn't that confusing? 
It might be useful in some situation but still we'll need a way to do a "clean" start, building containers over previously built / validated images, no matter what's in the current directory, won't we? 
By the way we also have the ports mapping from docker-compose.yml

**Comment by ptrovatelli** [comment]
there's something else we could try; see my comment in https://github.com/DefectDojo/django-DefectDojo/pull/1028#issuecomment-482000212

**Comment by aaronweaver** [comment]
Fixed

**Comment by petrabrunner** [comment]
@surewould thx for providing the workaround! 👍 I have been screening the logs for the origin of the problem but overlooked that line!

**Comment by ptrovatelli** [comment]
@surewould @petrabrunner if you feel like it, you're welcome to try out this PR and see how it works for you: https://github.com/DefectDojo/django-DefectDojo/pull/1045
It contains a few improvments on docker environment. You can look at the first comment to see what it's about and how to pull it, and DOCKER.md for how to use. We're waiting for feedback before merging it. Thanks :)

**Comment by petrabrunner** [comment]
@ptrovatelli I tested your pullrequest. do I still have to copy the dojo/settings/settings.dist.py to dojo/settings/settings.py or should that already be done by the code you changed? (because I noticed that it does not.)
